src/core/int128.d

```diff
@@ -0,0 +1,748 @@
+    Cent quotient;
+    while (shifts--)
+    {
+        //printf("shifts %d c1(%llx,%llx) c2(%llx,%llx)\n", shifts, c1.lo, c1.hi, c2.lo, c2.hi);
+        quotient = shl1(quotient);
+        if (uge(c1, c2))
+        {
+            //printf("sub\n");
+            c1 = sub(c1, c2);
+            quotient.lo |= 1;
+        }
+        c2 = shr1(c2);
+    }
+    modulus = c1;
+    //printf("quotient "); print(quotient);
+    //printf("modulus  "); print(modulus);
+    return quotient;
+}
+
+
+/****************************
+ * Signed divide c1 / c2.
+ * Params:
+ *      c1 = dividend
+ *      c2 = divisor
+ * Returns:
+ *      quotient c1 / c2
+ */
+pure
+Cent div(Cent c1, Cent c2)
+{
+    Cent modulus;
+    return divmod(c1, c2, modulus);
+}
+
+/****************************
+ * Signed divide c1 / c2. The remainder after division is stored to modulus.
+ * Params:
+ *      c1 = dividend
+ *      c2 = divisor
+ *      modulus = set to c1 % c2
+ * Returns:
+ *      quotient c1 / c2
+ */
+pure
+Cent divmod(Cent c1, Cent c2, out Cent modulus)
+{
+    /* Muck about with the signs so we can use the unsigned divide
+     */
+    if (cast(I)c1.hi < 0)
+    {
+        if (cast(I)c2.hi < 0)
+        {
+            Cent r = udivmod(neg(c1), neg(c2), modulus);
+            modulus = neg(modulus);
+            return r;
+        }
+        Cent r = neg(udivmod(neg(c1), c2, modulus));
+        modulus = neg(modulus);
+        return r;
+    }
+    else if (cast(I)c2.hi < 0)
+    {
+        return neg(udivmod(c1, neg(c2), modulus));
+    }
+    else
+        return udivmod(c1, c2, modulus);
+}
+
+/****************************
+ * If c1 > c2 unsigned
+ * Params:
+ *      c1 = operand 1
+ *      c2 = operand 2
+ * Returns:
+ *      true if c1 > c2
+ */
+pure
+bool ugt(Cent c1, Cent c2)
+{
+    return (c1.hi == c2.hi) ? (c1.lo > c2.lo) : (c1.hi > c2.hi);
+}
+
+/****************************
+ * If c1 >= c2 unsigned
+ * Params:
+ *      c1 = operand 1
+ *      c2 = operand 2
+ * Returns:
+ *      true if c1 >= c2
+ */
+pure
+bool uge(Cent c1, Cent c2)
+{
+    return !ugt(c2, c1);
+}
+
+/****************************
+ * If c1 < c2 unsigned
+ * Params:
+ *      c1 = operand 1
+ *      c2 = operand 2
+ * Returns:
+ *      true if c1 < c2
+ */
+pure
+bool ult(Cent c1, Cent c2)
+{
+    return ugt(c2, c1);
+}
+
+/****************************
+ * If c1 <= c2 unsigned
+ * Params:
+ *      c1 = operand 1
+ *      c2 = operand 2
+ * Returns:
+ *      true if c1 <= c2
+ */
+pure
+bool ule(Cent c1, Cent c2)
+{
+    return !ugt(c1, c2);
+}
+
+/****************************
+ * If c1 > c2 signed
+ * Params:
+ *      c1 = operand 1
+ *      c2 = operand 2
+ * Returns:
+ *      true if c1 > c2
+ */
+pure
+bool gt(Cent c1, Cent c2)
+{
+    if (cast(I)c1.hi >= 0)
+    {
+        if (cast(I)c2.hi >= 0)
+            return ugt(c1, c2);
+        return true;
+    }
+    if (cast(I)c2.hi >= 0)
+        return false;
+    return ugt(c1, c2);
+
```

Comment:
```suggestion
    return (c1.hi == c2.hi) ? (c1.lo > c2.lo) : (cast(I)c1.hi > cast(I)c2.hi);
```

Reply:
I am all in favor of these improvements. But what would really help is to approve this PR. That way, I can begin incorporating cent support into the compiler, while you can submit PRs for these optimizations. That way, you get the credit for the optimizations, and we can be working in parallel. Currently, not having this PR pulled is a blocker.

Reply:
As noted, I have nothing further to add after these remaining corrections.

Reply:
The logic I wrote is easier to understand and visually determine to be correct. That's why I prefer it as the baseline implementation. Your version is faster, but is hard to read, and so should go in a followup PR. That way, it the PR breaks things, we know where to look.